libxrdp/xrdp_sec.c

```diff
@@ -374,9 +374,47 @@ xrdp_load_keyboard_layout(struct xrdp_client_info *client_info)
     char keyboard_cfg_file[256] = { 0 };
     char rdp_layout[256] = { 0 };
 
-    LOG(LOG_LEVEL_INFO, "xrdp_load_keyboard_layout: keyboard_type [%d] keyboard_subtype [%d]",
-        client_info->keyboard_type, client_info->keyboard_subtype);
+    struct xrdp_keyboard_overrides *ko =
+            &client_info->xrdp_keyboard_overrides;
 
+    LOG(LOG_LEVEL_INFO, "xrdp_load_keyboard_layout: Keyboard information sent"
+        " by the RDP client, keyboard_type:[0x%02X], keyboard_subtype:[0x%02X],"
+        " keylayout:[0x%08X]",
+        client_info->keyboard_type, client_info->keyboard_subtype,
+        client_info->keylayout);
+
+    /* Store original keyboard information */
+    ko->original_type = client_info->keyboard_type;
+    ko->original_subtype = client_info->keyboard_subtype;
+    ko->original_layout = client_info->keylayout;
+
+    if (ko->type != 0)
+    {
+        LOG(LOG_LEVEL_DEBUG, "overrode keyboard_type 0x%02X"
+            " with 0x%02X", client_info->keyboard_type, ko->type);
+        client_info->keyboard_type = ko->type;
+    }
+    if (ko->subtype != 0)
+    {
+        LOG(LOG_LEVEL_DEBUG, "overrode keyboard_subtype 0x%02X"
+            " with 0x%02X", client_info->keyboard_subtype,
+            ko->subtype);
+        client_info->keyboard_subtype = ko->subtype;
+    }
+    if (ko->layout != 0)
+    {
+        LOG(LOG_LEVEL_DEBUG, "overrode keylayout 0x%08X"
+            " with 0x%08X", client_info->keylayout, ko->layout);
+        client_info->keylayout = ko->layout;
+    }
+    if ((ko->type | ko->subtype | ko->layout) != 0)
+    {
+        LOG(LOG_LEVEL_INFO, "xrdp_load_keyboard_layout: Overrode keyboard"
+            " infomation, keyboard_type:[0x%02X], keyboard_subtype:[0x%02X],"
+            " keylayout:[0x%08X]",
+            client_info->keyboard_type, client_info->keyboard_subtype,
+            client_info->keylayout);
+    }
```

Comment:
Suggest removing 410-417 entirely, and making the preceding messages `LOG_LEVEL_INFO`. You've already logged the client info at line 380-384. The override messages should also be at `LOG_LEVEL_INFO` level so we get a complete picture of the keyboard layout variables when running at `LOG_LEVEL_INFO`.

Otherwise we're just duplicating information when running at `LOG_LEVEL_DEBUG`.

Reply:
Thank you.
I was adding this for future code additions.
I've changed my mind and will remove lines 410-417.

I will change the override message to LOG_LEVEL_INFO.